lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -699,6 +700,93 @@ FIRRTLType MemOp::getDataTypeOrNull() {
 // Statements
 //===----------------------------------------------------------------------===//
 
+static bool areConnectEquivalent(FIRRTLType destType, FIRRTLType srcType);
+
+/// Helper to implement the equivalence logic for a pair of bundle elements in a
+/// connect.
+static bool areElementsConnectEquivalent(
+    std::tuple<BundleType::BundleElement, BundleType::BundleElement> tuple) {
+  BundleType::BundleElement destElement = std::get<0>(tuple);
+  Identifier destElementName = std::get<0>(destElement);
+  FIRRTLType destElementType = std::get<1>(destElement);
+  bool destIsFlip = destElementType.isa<FlipType>();
+
+  BundleType::BundleElement srcElement = std::get<1>(tuple);
+  Identifier srcElementName = std::get<0>(srcElement);
+  FIRRTLType srcElementType = std::get<1>(srcElement);
+  bool srcIsFlip = srcElementType.isa<FlipType>();
+
+  if (destElementName != srcElementName)
+    return false;
+
+  if (destIsFlip != srcIsFlip)
+    return false;
+
+  // If either destination or source is flipped, strip the outer flip for
+  // recursively equivalance checking.
+  if (destIsFlip)
+    destElementType = destElementType.cast<FlipType>().getElementType();
+
+  if (srcIsFlip)
+    srcElementType = srcElementType.cast<FlipType>().getElementType();
+
+  return areConnectEquivalent(destElementType, srcElementType);
+}
+
+/// Returns whether the two types are equivalent for the purpose of
+/// connections. See Section 4.5 in the FIRRTL spec:
```

Comment:
I'd avoid referencing a section number (which can drift) and URL, you can just mention the FIRRTL spec by name.